### PR TITLE
feat: allow reuse of persistent license sessions

### PR DESCRIPTION
**Pull request by valotvince**
Add capability to re-use persistent license sessions across sessions.

DrmEngine will now always:
- try to start stored persistent sessions before trying to fetch a license, as-to be able to check if all needed keys are already loaded.
- ask for a new license when the persistent session doesn't have the needed keys for playback,

Given the flag `persistentSessionOnlinePlayback` is true, DrmEngine:
- won't remove the persistent session from the device at the end of the playback,
- won't throw an error when the persistent session isn't found on the device,

For now, it needs Shaka's users to persist session information by themselves (localStorage, IndexDB, ...) before giving it back for the next session. Still, it lays foundation to develop the feature to fully handling it on Shaka's side.

TBD:
- [x] Rename all `offlineSessionIds_` in DrmEngine to `storedPersistentSessionIds_` because we now use those in two different contexts, online / offline (and in comments too) (after full review to avoid noise over the feature)
- [ ] Rename `loadOfflineSession_` in DrmEngine to `loadStoredPersistentSession_` (after full review to avoid noise over the feature)
- [x] Add a public API to retrieve the session (ids and expiration) from DRM Engine if not already available
- [x] Refactor / Set up another way to give persistentSessionIds to also give associated initData with it
- [x] Test full offline to ensure no regressions
- [x] Document feature usage

Related to #1956

### CODE REVIEW

File: test/media/drm_engine_unit.js [hunk 1]
```diff
@@ -1087,6 +1088,149 @@ describe('DrmEngine', () => {
           shaka.util.Error.Code.FAILED_TO_GENERATE_LICENSE_REQUEST,
           message, nativeError, undefined));
     });
+
+    // it('should throw a OFFLINE_SESSION_REMOVED error', async () => {
```
Comment (valotvince) [inline test/media/drm_engine_unit.js:1092]: Because the error makes the createOrLoad throw, I am unable to set up that test (because of the unhandled promise rejection).

I could I set up the test correctly ? 

I've thought to noop createOrLoad in the attach method, because the errors there should have been handled through the onError_method...
```
createOrLoad.catch(() => {});
```
I've tried to noop the catch of allSessionsLoaded_ without success (and it was already nooped in constructor)

Comment (valotvince) [reply test/media/drm_engine_unit.js]: Turns out it was because I've changed the createOrLoad signature to an async one, errors weren't caught before, so I've silence them

**Comment by avelad** [comment]
Related to https://github.com/shaka-project/shaka-player/issues/1956

File: lib/media/drm_engine.js [hunk 2]
```diff
@@ -617,23 +636,40 @@ shaka.media.DrmEngine = class {
    *
    * @return {!Promise}
    */
-  createOrLoad() {
-    // Create temp sessions.
+  async createOrLoad() {
```
Comment (caridley) [inline lib/media/drm_engine.js:639]: I noticed that the attach(video) function calls createOrLoad() but does not wait for the returned promise to resolve. Should it wait?

Comment (valotvince) [reply lib/media/drm_engine.js:639]: I think it was done on purpose so that creating / loading sessions could be done in parallel of fetching manifest information, segments and more.

File: externs/shaka/manifest.js [hunk 3]
```diff
@@ -171,7 +171,6 @@ shaka.extern.InitDataOverride;
  */
 shaka.extern.DrmInfo;
 
-
```
Comment (avelad) [inline externs/shaka/manifest.js:174]: Please revert this change

**Review by avelad — CHANGES_REQUESTED** [review]
In general, LGTM, but I think we should give the possibility (player configuration) that ShakaPlayer itself stores this information in localStorage for example.

**Comment by avelad** [comment]
@valotvince it seems that there are some conflict with the main branch, can you rebase? Thanks!

**Comment by valotvince** [comment]
👋 @avelad @joeyparrish I've rebased on main & added documentation about the feature, could you review it again ? Thanks :)

> I think we should give the possibility (player configuration) that ShakaPlayer itself stores this information in localStorage for example.
If that's possible I'd like it to be done in another PR, as localStorage (or IndexeDB) is not necessarily the best on some devices. It might take more reflexion on the whole storage capabilities of Shaka, like allowing to give a StorageManager like the ABRManager which would handle the storage, without Shaka's knowing what the storage is.

File: lib/media/drm_engine.js [hunk 4]
```diff
@@ -1185,16 +1218,18 @@ shaka.media.DrmEngine = class {
...
+          this.onError_(new shaka.util.Error(
+              shaka.util.Error.Severity.CRITICAL,
+              shaka.util.Error.Category.DRM,
+              shaka.util.Error.Code.OFFLINE_SESSION_REMOVED));
+
+          return Promise.resolve();
+        }
       }
 
-      // TODO: We should get a key status change event.  Remove once Chrome CDM
```
[the remark this answers is not in the document]
Comment (valotvince) [reply lib/media/drm_engine.js:1217]: Downloaded Widevine DRM content worked great with Chrome ;)

[the remark this answers is not in the document]
Comment (avelad) [reply lib/media/drm_engine.js:1217]: What happens in old TV browsers?

[the remark this answers is not in the document]
Comment (valotvince) [reply lib/media/drm_engine.js:1217]: Haven't tested yet the persistent licenses on TV but will do

[the remark this answers is not in the document]
Comment (valotvince) [reply lib/media/drm_engine.js:1217]: ### Tizen 6.5
It correctly starts the persistent-license session at first launch but fails on the session load at second launch, so my guess is that this isn't supported.
The sessionId is always the same `initialdrmsessionid` so I don't think it could work at all.
```
Shaka Error DRM.FAILED_TO_CREATE_SESSION (Failed to execute 'load' on 'MediaKeySession': Loading session failed. . (2))
```

### webOS 5.0
It fails to request a media key system access when asking for a `persistent-license` with the error `REQUESTED_KEY_SYSTEM_CONFIG_UNAVAILABLE`, so it seems it doesn't support persistent license at all


<br />

So my overall conclusion is persistent license isn't supported on TV devices, so I don't think we need to go further into making this feature working on those devices.


<br />

### Error Recoverability
It pointed me out on an issue with my code where fails to load a persistent license will stop the player, rather than trying to fetch a new one (intended behaviour).


I have two options for that, let me know what you prefer (my pref is option 1):

- Option 1: Send a RECOVERABLE error
```
this.onError_(new shaka.util.Error(
    this.config_.persistentSessionOnlinePlayback ?
      shaka.util.Error.Severity.RECOVERABLE : shaka.util.Error.Severity.CRITICAL;
    shaka.util.Error.Category.DRM,
    shaka.util.Error.Code.FAILED_TO_CREATE_SESSION,
    error.message));
```

- Option 2: Bypass the error, only log something:
```
if (this.config_.persistentSessionOnlinePlayback) {
  shaka.log.warning('Persistent session', sessionId, 'not available');
} else {
  this.onError_(new shaka.util.Error(
      shaka.util.Error.Severity.CRITICAL,
      shaka.util.Error.Category.DRM,
      shaka.util.Error.Code.OFFLINE_SESSION_REMOVED));

  return Promise.resolve();
}
```

**Comment by valotvince** [comment]
Hi @avelad @joeyparrish, any comments to share on this PR ? Thanks !

File: lib/media/drm_engine.js [hunk 5]
```diff
@@ -1207,32 +1297,42 @@ shaka.media.DrmEngine = class {
...
             shaka.util.Error.Code.OFFLINE_SESSION_REMOVED));
-        return Promise.resolve();
-      }
 
-      // TODO: We should get a key status change event.  Remove once Chrome CDM
-      // is fixed.
-      metadata.loaded = true;
-      if (this.areAllSessionsLoaded_()) {
-        this.allSessionsLoaded_.resolve();
+        metadata.loaded = true;
```
Comment (theodab) [inline lib/media/drm_engine.js:1309]: Do we really need to set `metadata.loaded` to `true` here?
All that value does is determine when `areAllSessionsLoaded_` returns `true`. So if we are manually setting this to true for each stream we call `loadOfflineSession_` on, and we call `createOrLoad` on each session at once inside `createOrLoad`, doesn't that mean that we will just immediately resolve `this.allSessionsLoaded_` when the last session begins to load? That seems like it defeats the purpose of waiting for the key status change event, if we're still resolving before the event fires.

File: test/media/drm_engine_unit.js [hunk 6]
```diff
@@ -1093,6 +1093,178 @@ describe('DrmEngine', () => {
...
+
+      expect(mockMediaKeys.createSession).toHaveBeenCalledTimes(1);
+      expect(mockMediaKeys.createSession)
+          .toHaveBeenCalledWith('persistent-license');
+      expect(session1.load).toHaveBeenCalledWith('persistent-session-id');
+
+      expect(session2.generateRequest).not.toHaveBeenCalled();
+    });
+
+    it(
```
Comment (theodab) [inline test/media/drm_engine_unit.js:1173]: I think this would fit into one line. No need to do the weird indenting.

**Review by theodab — DISMISSED** [review]

File: externs/shaka/player.js [hunk 7]
```diff
@@ -680,12 +680,60 @@ shaka.extern.ProducerReferenceTime;
...
+/**
+ * @typedef {{
+ *   sessionId: string,
+ *   sessionType: string,
+ *   initData: ?Uint8Array,
+ *   initDataType: ?string
+ * }}
+ *
+ * @description
+ * DRM Session Metadata for as session
```
[the remark this answers is not in the document]
Comment (valotvince) [reply externs/shaka/player.js:692]: Went for:
> DRM Session Metadata

File: externs/shaka/player.js [hunk 8]
```diff
@@ -680,12 +680,60 @@ shaka.extern.ProducerReferenceTime;
...
+
+/**
+ * @typedef {{
+ *   sessionId: string,
+ *   initData: ?Uint8Array,
+ *   initDataType: ?string
+ * }}
+ *
+ * @description
+ * DRM Session Metadata for as session
```
[the remark this answers is not in the document]
Comment (valotvince) [reply externs/shaka/player.js:715]: You are correct ! :)

File: lib/media/drm_engine.js [hunk 9]
```diff
@@ -1207,32 +1297,42 @@ shaka.media.DrmEngine = class {
...
 
+        const severity = this.config_.persistentSessionOnlinePlayback ?
+            shaka.util.Error.Severity.RECOVERABLE :
+            shaka.util.Error.Severity.CRITICAL;
+
         this.onError_(new shaka.util.Error(
-            shaka.util.Error.Severity.CRITICAL,
+            severity,
             shaka.util.Error.Category.DRM,
             shaka.util.Error.Code.OFFLINE_SESSION_REMOVED));
```
[the remark this answers is not in the document]
Comment (valotvince) [reply lib/media/drm_engine.js:1307]: I'll change the description of the error to:
> A required offline session was removed. The content might not be playable depending of the playback context.

File: lib/media/drm_engine.js (hunk 5 above)
Comment (valotvince) [reply lib/media/drm_engine.js:1309]: We only set `metadata.loaded` when the session is not present anymore (we will never receive a `keystatuschange` event), and send a `OFFLINE_SESSION_REMOVED` event.

If other sessions are present & correctly loaded, we wait for the `keystatuschange` to set the session as loaded.

File: test/media/drm_engine_unit.js (hunk 6 above)
Comment (valotvince) [reply test/media/drm_engine_unit.js:1173]: The whole line
```
it('tries persistent session ids before requesting a license', async () => {
```
is 81 characters long 😞

**Review by avelad — CHANGES_REQUESTED** [review]
LGTM, can you rebase and resolve the conflicts? Thanks!

**Comment by valotvince** [comment]
@avelad 👋  Rebase done !

**Review by avelad — APPROVED** [review]

File: docs/tutorials/drm-config.md [hunk 10]
```diff
@@ -208,3 +208,64 @@ you should provide an empty string as robustness
...
+    persistentSessionsMetadata: [{
+      sessionId: 'deadbeefdeadbeefdeadbeefdeadbeef',
+      initData: new InitData(0),
+      initDataType: 'cenc'
+    }]
+  }
+});
+```
+
+NB: Shaka doesn't provide a out-of-the-box storage mechanism for the sessions
```
Comment (joeyparrish) [inline docs/tutorials/drm-config.md:265]: There is such a mechanism in the offline/ codebase.  We could discuss a follow-up PR to use that mechanism if/when offline support is included.

**Review by joeyparrish — APPROVED** [review]

**Comment by joeyparrish** [comment]
Failed tests look like unrelated flakiness, but I'm re-running them just to be sure.
